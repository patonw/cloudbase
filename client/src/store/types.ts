export type UUID = string

export interface Workbook {
  uuid: UUID,
  name: string,
  sheets: UUID[],
}

export interface Worksheet {
  uuid: UUID,
  workbook: UUID,
  name: string,
  cells: UUID[],
  modified?: number,
  process?: UUID,
}

export interface CellResult {
  progress?: number
  data?: string
  json?: () => any
  error?: any
}

export interface CodeCell {
  uuid: UUID,
  worksheet: UUID,
  lang: string,
  script: string,
}

export interface GraphCell extends CodeCell {
  spec: string
}

export type Cell = CodeCell | GraphCell

export function isGraphCell(cell: Cell): cell is GraphCell {
  return (cell as GraphCell).spec !== undefined
}

