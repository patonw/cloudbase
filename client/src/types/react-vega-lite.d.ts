declare module 'react-vega-lite' {
  export function createClassFromLiteSpec(...args: any[]): any;
  export default function _default(a: any): any;
}