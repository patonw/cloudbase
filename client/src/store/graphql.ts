import { ajax, AjaxResponse } from 'rxjs/ajax';
import { Observable } from 'rxjs';
import { UUID, Cell, isGraphCell } from './types'
import { LoadWorksheetActionData } from './actions';

type Ajax = typeof ajax

export interface GraphQLRequest {
  query: string
  variables?: {
    [key: string]: any
  }
}

export type GraphQLClient = (body: GraphQLRequest) => Observable<AjaxResponse>

export function makeAjaxClient(ajax: Ajax): GraphQLClient {
  return (body: GraphQLRequest) => ajax({
    url: '/graphql',
    responseType: 'json',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'rxjs-custom-header': 'Rxjs'
    },
    body
  })
}

export interface GraphQLResponse {
  errors?: any
}

export interface LoadTOCResponse extends GraphQLResponse {
  data: {
    allWorkbooks: {
      uuid: UUID
      name: string
      sheets: {
        uuid: UUID
        name: string
      }[]
    }[],

    allProcesses: {
      uuid: UUID
      sheet: {
        uuid: UUID
      }
    }[],
  }
}

export const loadTOC = () => ({
  query: `
  query loadTOC {
    allWorkbooks {
      uuid
      name
      sheets {
        uuid
        name
      }
    }

    allProcesses {
      uuid
      sheet {
        uuid
      }
    }
  }
`})

export interface LoadWorksheetResponse extends GraphQLResponse {
  data: {
    worksheet: LoadWorksheetActionData
  }
}

export const loadWorksheet = (uuid: UUID) => ({
  query: `
  query loadSheet($sheetId: ID!) {
    worksheet(sheetId: $sheetId) {
      uuid
      name
      cells {
        __typename
        uuid
        script
        ... on GraphCell {
          spec
        }
      }
    }
  }`,
  variables: {
    sheetId: uuid
  }
})

export interface ExecuteCellResponse extends GraphQLResponse {
  data: {
    setCellScript?: {
      uuid: UUID
    },
    executeCell: {
      cell: {
        uuid: UUID
      }
      data: string
      json: string
    }
  }
}
export const updateAndExecute = (processId: UUID, cellId: UUID, dirty: Cell) => {
  let updates = ""
  let args = ""
  let variables: any = {
    processId,
    cellId,
  }

  if (dirty.script) {
    updates += `
      setCellScript(cellId: $cellId, script: $script) {
        uuid
      }`
    args += ', $script: String!'
    variables.script = dirty.script
  }

  if (isGraphCell(dirty)) {
    updates += `
      setGraphSpec(cellId: $cellId, spec: $spec) {
        uuid
      }
    `
    args += ', $spec: String!'
    variables.spec = dirty.spec
  }

  return ({
    query: `
    mutation updateAndExecute($processId:ID!, $cellId:ID! ${args}) {
      ${updates}
      executeCell(processId: $processId, cellId: $cellId) {
        cell {
          uuid
        }
        data
        json
      }
    }`,
    variables,
  })
}

export const executeCell = (processId: UUID, cellId: UUID, dirty?: Cell) => (dirty? updateAndExecute(processId, cellId, dirty) : {
  query: `
  mutation justExecute($processId:ID!, $cellId:ID!) {
    executeCell(processId: $processId, cellId: $cellId) {
      cell {
        uuid
      }
      data
      json
    }
  }`,
  variables: {
    processId,
    cellId,
  }
})