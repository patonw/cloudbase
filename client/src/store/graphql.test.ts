import * as gql from './graphql'
import { CodeCell, GraphCell } from './types'

it('is all good', () => {
  1+ 1
})

it('generates a TOC query', () => {
  const query = gql.loadTOC()
  expect(query).toMatchSnapshot()
})

it('generates a worksheet query', () => {
  const query = gql.loadWorksheet('helloWorld')
  expect(query).toMatchSnapshot()
})

it('generates a combined update and execute mutation for a code cell', () => {
  const dirty: CodeCell =  {
    uuid: 'theCellId',
    lang: 'kotlin',
    worksheet: 'theWorksheetId',
    script: `1+1`,
  }
  const query = gql.executeCell('thePID', 'theCellId', dirty)
  expect(query).toMatchSnapshot()
})


it('generates a combined update and execute mutation for a graph cell', () => {
  const dirty: GraphCell =  {
    uuid: 'theCellId',
    lang: 'kotlin',
    worksheet: 'theWorksheetId',
    script: `1+1`,
    spec: `
    {
      "$schema": "https://vega.github.io/schema/vega-lite/v3.json",
      "description": "A simple bar chart with embedded data.",
      "mark": "bar",
      "encoding": {
        "x": {"field": "first", "type": "ordinal"},
        "y": {"field": "second", "type": "quantitative"}
      }
    }`
  }
  const query = gql.executeCell('thePID', 'theCellId', dirty)
  expect(query).toMatchSnapshot()
})

it('only executes a clean cell without updating', () => {
  const query = gql.executeCell('thePID', 'theCellID')
  expect(query).toMatchSnapshot()
})

it('inserts a code cell', () => {
  const query = gql.insertCell('theCellID', undefined, 'CODE')
  expect(query).toMatchSnapshot()
})

it('inserts a graph cell', () => {
  const query = gql.insertGraphCell('theCellID', 5)
  expect(query).toMatchSnapshot()
})

it('deletes a cell', () => {
  const query = gql.deleteCell('theSheetID', 'theCellID')
  expect(query).toMatchSnapshot()
})

it('creates a new worksheet', () => {
  const query = gql.createWorksheet('theWorkbook', "Hello Worksheet")
  expect(query).toMatchSnapshot()
})

it('creates a new process', () => {
  const query = gql.createProcess('theWorksheet')
  expect(query).toMatchSnapshot()
})