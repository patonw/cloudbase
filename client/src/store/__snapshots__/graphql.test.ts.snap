// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a new process 1`] = `
Object {
  "query": "
      mutation CreateProcess($sheetId: ID!) {
        createProcess(sheetId: $sheetId) {
          uuid
          sheet {
            uuid
          }
        }
      }
    ",
  "variables": Object {
    "sheetId": "theWorksheet",
  },
}
`;

exports[`creates a new worksheet 1`] = `
Object {
  "query": "
      mutation CreateWorksheet($bookId: ID!, $name: String) {
        createWorksheet(workbookId: $bookId, name: $name) {
          uuid
          name
          cells {
            __typename
            uuid
            script
            ... on GraphCell {
              spec
            }
          }
        }
      }
    ",
  "variables": Object {
    "bookId": "theWorkbook",
    "name": "Hello Worksheet",
  },
}
`;

exports[`deletes a cell 1`] = `
Object {
  "query": "
      mutation Apoptosis($sheetId: ID!, $cellId: ID!) {
        deleteCell(sheetId: $sheetId, cellId: $cellId) {
          sheet {
            uuid
          }
          cell {
            uuid
          }
        }
      }
    ",
  "variables": Object {
    "cellId": "theCellID",
    "sheetId": "theSheetID",
  },
}
`;

exports[`generates a TOC query 1`] = `
Object {
  "query": "
  query loadTOC {
    allWorkbooks {
      uuid
      name
      sheets {
        uuid
        name
      }
    }

    allProcesses {
      uuid
      sheet {
        uuid
      }
    }
  }
",
}
`;

exports[`generates a combined update and execute mutation for a code cell 1`] = `
Object {
  "query": "
    mutation updateAndExecute($processId:ID!, $cellId:ID! , $script: String!) {
      
      setCellScript(cellId: $cellId, script: $script) {
        uuid
      }
      executeCell(processId: $processId, cellId: $cellId) {
        cell {
          uuid
        }
        data
        json
      }
    }",
  "variables": Object {
    "cellId": "theCellId",
    "processId": "thePID",
    "script": "1+1",
  },
}
`;

exports[`generates a combined update and execute mutation for a graph cell 1`] = `
Object {
  "query": "
    mutation updateAndExecute($processId:ID!, $cellId:ID! , $script: String!, $spec: String!) {
      
      setCellScript(cellId: $cellId, script: $script) {
        uuid
      }
      setGraphSpec(cellId: $cellId, spec: $spec) {
        uuid
      }
      executeCell(processId: $processId, cellId: $cellId) {
        cell {
          uuid
        }
        data
        json
      }
    }",
  "variables": Object {
    "cellId": "theCellId",
    "processId": "thePID",
    "script": "1+1",
    "spec": "
    {
      \\"$schema\\": \\"https://vega.github.io/schema/vega-lite/v3.json\\",
      \\"description\\": \\"A simple bar chart with embedded data.\\",
      \\"mark\\": \\"bar\\",
      \\"encoding\\": {
        \\"x\\": {\\"field\\": \\"first\\", \\"type\\": \\"ordinal\\"},
        \\"y\\": {\\"field\\": \\"second\\", \\"type\\": \\"quantitative\\"}
      }
    }",
  },
}
`;

exports[`generates a worksheet query 1`] = `
Object {
  "query": "
  query loadSheet($sheetId: ID!) {
    worksheet(sheetId: $sheetId) {
      uuid
      name
      cells {
        __typename
        uuid
        script
        ... on GraphCell {
          spec
        }
      }
    }
  }",
  "variables": Object {
    "sheetId": "helloWorld",
  },
}
`;

exports[`inserts a code cell 1`] = `
Object {
  "query": "
      mutation InsertCell($sheetId: ID!, $index: Int, $cellType: CellType) {
        insertCell(sheetId: $sheetId, index: $index, cellType: $cellType) {
          sheet {
            uuid
            cells {
              uuid
            }
          }
          cell {
            __typename
            uuid
            script
            ... on GraphCell {
              spec
            }
          }
        }
      }
    ",
  "variables": Object {
    "cellType": "CODE",
    "sheetId": "theCellID",
  },
}
`;

exports[`inserts a graph cell 1`] = `
Object {
  "query": "
      mutation InsertCell($sheetId: ID!, $index: Int, $cellType: CellType) {
        insertCell(sheetId: $sheetId, index: $index, cellType: $cellType) {
          sheet {
            uuid
            cells {
              uuid
            }
          }
          cell {
            __typename
            uuid
            script
            ... on GraphCell {
              spec
            }
          }
        }
      }
    ",
  "variables": Object {
    "cellType": "GRAPH",
    "index": 5,
    "sheetId": "theCellID",
  },
}
`;

exports[`only executes a clean cell without updating 1`] = `
Object {
  "query": "
  mutation justExecute($processId:ID!, $cellId:ID!) {
    executeCell(processId: $processId, cellId: $cellId) {
      cell {
        uuid
      }
      data
      json
    }
  }",
  "variables": Object {
    "cellId": "theCellID",
    "processId": "thePID",
  },
}
`;
